File: Exercism.io/python/meetup/meetup.py

import datetime
import calendar

DAYRULES = { 'teenth': lambda d: [dn for dn in d if dn in range(13,20)][0],
             '1st':    lambda d: d[0],
             '2nd':    lambda d: d[1],
             '3rd':    lambda d: d[2],
             '4th':    lambda d: d[3],
             'last':   lambda d: d[-1] }

def meetup_day(year, month, day, rule):
    try:
        days = calendar.Calendar().itermonthdays2(year, month)
        daycode = list(calendar.day_name).index(day)
        possible_days = [ dn for (dn, dc) in days if dn > 0 and dc == daycode ]
        resulting_day = DAYRULES[rule](possible_days)
        return datetime.date(year, month, resulting_day)
    except:
        return None

File: Exercism.io/elixir/meetup/meetup.exs
defmodule Meetup do

  @days [monday: 1, tuesday: 2, wednesday: 3, 
         thursday: 4, friday: 5, saturday: 6, 
         sunday: 7]

  def meetup(year, month, weekday, schedule) do
    daycode = @days[weekday]
    1..(:calendar.last_day_of_the_month(year, month))
      |> Enum.filter(&(right_weekday?(year, month, &1, daycode)))
      |> scheduled_day(schedule)
      |> to_datetuple(year, month)
  end

  defp right_weekday?(year, month, day, daycode) do
    :calendar.day_of_the_week(year, month, day) == daycode
  end

  defp scheduled_day(days, :teenth), do: Enum.find(days, &(&1 >= 13))
  defp scheduled_day(days, :first),  do: Enum.at(days, 0)
  defp scheduled_day(days, :second), do: Enum.at(days, 1)
  defp scheduled_day(days, :third),  do: Enum.at(days, 2)
  defp scheduled_day(days, :fourth), do: Enum.at(days, 3)
  defp scheduled_day(days, :last),   do: List.last(days)

  defp to_datetuple(day, year, month), do: {year, month, day}

end
